src/jdk/src/share/sample/jmx/heapmemorymonitor/HeapMemoryMonitor.java

```diff
@@ -0,0 +1,166 @@
+                continue;
+            }
+            heapMemoryPoolMXBeans.add(memoryPool);
+
+            memoryPoolName = memoryPool.getName();
+            if (memoryPoolName.contains("Eden")) {
+                // Ignore Eden, since it's just an allocation buffer
+                continue;
+            }
+            afterGCHeapMemoryPoolMXBeans.add(memoryPool);
+        }
+
+	df.setRoundingMode(RoundingMode.CEILING);
+    }
+
+    private void run() {
+        dump("                                 Start");
+        allocateObjects();
+        System.gc(); // Promote everything to old gen
+        dump("After allocation and old gen promotion");
+        allocateObjects();
+        dump("    After more allocation (split gens)");
+        System.gc();
+        dump("               After old gen promotion");
+    }
+
+    private void dump(String header) {
+        System.out.println(header + ": HeapMemory        " + getHeapMemory() + " = " +
+                           df.format(getHeapMemoryUse()) + '%');
+        System.out.println(header + ": HeapMemoryAfterGC " + getHeapMemoryAfterGC() + " = " +
+                           df.format(getHeapMemoryAfterGCUse()) + '%');
+        System.out.flush();
+    }
+
+    public String getHeapMemory() {
+        getMemoryUsage();
+        return usedKb + "/" + totalKb + " Kb";
+    }
+
+    public double getHeapMemoryUse() {
+        getMemoryUsage();
+        if (totalKb == 0) return 0;
+        return ((double)usedKb * 100) / (double)totalKb;
+    }
+
+    public String getHeapMemoryAfterGC() {
+        getAfterGCMemoryUsage();
+        return afterGCUsedKb + "/" + afterGCTotalKb + " Kb";
+    }
+
+    public double getHeapMemoryAfterGCUse() {
+        getAfterGCMemoryUsage();
+        if (afterGCTotalKb == 0) return 0;
+        return ((double)afterGCUsedKb * 100) / (double)afterGCTotalKb;
+    }
+
+    private void getMemoryUsage() {
+        // Report on the sum of HEAP spaces
+        long used, max;
+	MemoryUsage usage;
+        for (MemoryPoolMXBean memoryPool : heapMemoryPoolMXBeans) {
+	    //            System.out.println("getMemoryUsage: " + memoryPool.getName());
+            usage = memoryPool.getUsage();
+            used = usage.getUsed();
+            usedKb += used / 1024;
+            max = usage.getMax();
+            // max can be undefined (-1).
+            // See http://docs.oracle.com/javase/7/docs/api/java/lang/management/MemoryUsage.html
+            totalKb += max == -1 ? 0 : max / 1024;
+	    //	    System.out.println("Used " + usedKb + " Max " + totalKb);
+        }
+    }
+
+    private void getAfterGCMemoryUsage() {
+        // Report on the sum of non-Eden HEAP spaces after the last gc
+        long used, max;
+	MemoryUsage usage;
+        for (MemoryPoolMXBean memoryPool : afterGCHeapMemoryPoolMXBeans) {
+	    //            System.out.println("getAfterGCMemoryUsage: " + memoryPool.getName());
+            usage = memoryPool.getCollectionUsage();
```

Comment:
Declare here, not outside the loop. Ideally also make this variable "final".

Reply:
Declared outside the loop to avoid constructing a new object every time around the loop.

Reply:
But, I'm not actually creating a new object, so point taken.